client/rest/src/main/java/org/elasticsearch/client/ResponseListener.java

```diff
@@ -23,6 +23,10 @@
  * Listener to be provided when calling async performRequest methods provided by {@link RestClient}.
  * Those methods that do accept a listener will return immediately, execute asynchronously, and notify
  * the listener whenever the request yielded a response, or failed with an exception.
+ *
+ * <p>
+ * Note that it is <strong>not</strong> safe to call {@link RestClient#close()} from either of these
+ * callbacks.
```

Comment:
I wanted to add some test to the `RestClient` itself to prevent this but I'm not sure the complexity is worth it. We'd have to override the `ThreadFactory` and store the threads and check if we are in one of them when we call `close`, I think.